openquake/hazardlib/near_fault.py

```diff
@@ -0,0 +1,476 @@
+
+    p0 = np.array(p0_xyz)
+    p1 = np.array(p1_xyz)
+    p2 = np.array(p2_xyz)
+
+    u = p1 - p0
+    v = p2 - p0
+
+    # vector normal to plane, ax+by+cy = d, normal=(a,b,c)
+
+    normal = np.cross(u, v)
+
+    # Define the d for the plane equation
+    dist_to_plane = np.dot(p0, normal)
+
+    return normal, dist_to_plane
+
+
+def projection_pp(site, normal, dist_to_plane, reference):
+    '''
+    Find the projection point Pp on the patch plane
+
+    :param site:
+        Location of the site, [lon, lat, dep]
+    :param normal:
+        normal of the plane, describe by a normal vector[a, b, c]
+    :param dist_to_plane:
+            d in the plane equation,  ax + by + cz = d
+    :param reference:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of project reference point
+    :returns:
+        pp, the projection point, [ppx, ppy, ppz], in xyz domain
+        format in numpy array
+    '''
+
+    # Transform to xyz coordinate
+    [site_x, site_y, site_z] = get_xyz_from_ll(site, reference)
+
+    a = np.array([(1, 0, 0, -normal[0]),
+                  (0, 1, 0, -normal[1]),
+                  (0, 0, 1, -normal[2]),
+                  (normal[0], normal[1], normal[2], 0)])
+    b = np.array([site_x, site_y, site_z, dist_to_plane])
+
+    x = np.linalg.solve(a, b)
+    pp = np.array([x[0], x[1], x[2]])
+    return pp
+
+
+def vectors2angle(v1, v2):
+    """ Returns the angle in radians between vectors 'v1' and 'v2'.
+
+    :param v1:
+        vector, format in numpy array
+    :param v2:
+        vector, format in numpy array
+    :return:
+        the angle in radians between the two vetors
+    """
+    cosang = np.dot(v1, v2)
+    sinang = np.linalg.norm(np.cross(v1, v2))
+    return np.arctan2(sinang, cosang)
+
+
+def average_s_rad(site, hypocenter, reference, pp,
+                  normal, dist_to_plane, e, p0, p1, delta_slip):
+    """
+    Get the average S-wave radiation pattern over E-path desrcibed in:
+    "Final report of the NGA-West2 directivity working group." (2013) by
+    Spudich, Paul, et al. and publish in PEER report, page 90- 92. Also
+    obtain the site to direct point distance, rd, and the hypocentre
+    distance, r_hyp.
+
+    :param site:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of the target site
+    :param hypocenter:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of hypocenter
+    :param reference:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location
+        of the reference point for coordinate projection within the
+        calculation. The suggested reference point is Epicentre.
+    :param pp:
+        the projection point pp on the patch plane,
+        format in numpy array
+    :param normal:
+        normal of the plane, describe by a normal vector[a, b, c]
```

Comment:
Are a, b, c the direction cosines of the plane including the fault patch considered here?